presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java

```diff
@@ -88,6 +90,31 @@ public SerializedPage serialize(Page page)
         return new SerializedPage(slice, markers, page.getPositionCount(), uncompressedSize);
     }
 
+    public SerializedPage wrapBuffer(Slice buffer, int positionCount)
+    {
+        byte markers = PageCodecMarker.none();
+        if (!compressor.isPresent()) {
+            return new SerializedPage(buffer, markers, positionCount, buffer.length());
+        }
+
+        int maxCompressedLength = maxCompressedLength(buffer.length());
+        if (compressionBuffer == null || compressionBuffer.length < maxCompressedLength) {
+            compressionBuffer = new byte[maxCompressedLength];
+        }
+        int actualCompressedLength = compressor.get().compress((byte[]) buffer.getBase(), 0, buffer.length(), compressionBuffer, 0, maxCompressedLength);
+
+        if (((1.0 * actualCompressedLength) / buffer.length()) > MINIMUM_COMPRESSION_RATIO) {
```

Comment:
To improve readability, I'd make this method follow the existing `serialize` like so:

```
    public SerializedPage wrapBuffer(Slice slice, int positionCount)
    {
        if (!compressor.isPresent()) {
            return new SerializedPage(slice, PageCodecMarker.none(), positionCount, slice.length());
        }

        int uncompressedSize = slice.length();
        int maxCompressedSize = compressor.get().maxCompressedLength(uncompressedSize);
        ensureCompressionBufferCapacity(maxCompressedSize);
        int compressedSize = compressor.get().compress((byte[]) slice.getBase(), 0, uncompressedSize, compressionBuffer, 0, maxCompressedSize);

        if ((((double) compressedSize) / uncompressedSize) <= MINIMUM_COMPRESSION_RATIO) {
            return new SerializedPage(
                    Slices.copyOf(Slices.wrappedBuffer(compressionBuffer, 0, compressedSize)),
                    COMPRESSED.set(PageCodecMarker.none()),
                    positionCount,
                    uncompressedSize);
        }

        return new SerializedPage(slice, PageCodecMarker.none(), positionCount, uncompressedSize);
    }

    private void ensureCompressionBufferCapacity(int capacity)
    {
        if (compressionBuffer == null || compressionBuffer.length < capacity) {
            compressionBuffer = new byte[capacity];
        }
    }
```

Reply:
I refactored the two serialize method as follows:
```
    public SerializedPage serialize(Page page)
    {
        SliceOutput serializationBuffer = new DynamicSliceOutput(toIntExact(page.getSizeInBytes() + Integer.BYTES)); // block length is an int
        writeRawPage(page, serializationBuffer, blockEncodingSerde);
        Slice slice = serializationBuffer.slice();
        return serialize(slice, page.getPositionCount());
    }

    public SerializedPage serialize(Slice slice, int positionCount)
    {
        int uncompressedSize = slice.length();
        byte markers = PageCodecMarker.none();

        if (compressor.isPresent()) {
            ByteBuffer compressionBuffer = ByteBuffer.allocate(compressor.get().maxCompressedLength(uncompressedSize));
            compressor.get().compress(slice.toByteBuffer(), compressionBuffer);
            compressionBuffer.flip();
            if ((((double) compressionBuffer.remaining()) / uncompressedSize) <= MINIMUM_COMPRESSION_RATIO) {
                slice = Slices.wrappedBuffer(compressionBuffer);
                markers = COMPRESSED.set(markers);
            }
        }

        if (spillCipher.isPresent()) {
            slice = Slices.wrappedBuffer(spillCipher.get().encrypt(slice.toByteBuffer()));
            markers = ENCRYPTED.set(markers);
        }
        else {
            //  Encryption disabled, slice data is likely much smaller than its backing buffer
            //  either because of compression or dynamic sizing of the initial output slice
            slice = Slices.copyOf(slice);
        }

        return new SerializedPage(slice, markers, positionCount, uncompressedSize);
    }
```